compose/config/config.py

```diff
@@ -1030,7 +1030,13 @@ def resolve_volume_paths(working_dir, service_dict):
 
 
 def resolve_volume_path(working_dir, volume):
-    container_path, host_path = split_path_mapping(volume)
+    if isinstance(volume, dict):
+        host_path = volume.get('source')
+        container_path = volume.get('target')
+        if host_path and volume.get('read_only'):
+            container_path += ':ro'
```

Comment:
I think there is also `no-copy`(from volume.nocopy) and the mount propagation flags that would need to be appended here.

It might be safer to convert everything to a dict, instead of converting to a string?